Simplify overbright implementation

slipher opened the pull request:
AIUI overbright simply means that the lightmaps and light grid should be scaled from 0 to 4, instead of 0 to 1, so we can represent a larger range of brightness. But image colors are only allowed to go from 0 to 1. This is a big problem... if we are using GL 1 (hence, perhaps, the mysterious "hardware implementations"). But we have shaders, so all we have to do is multiply by 4 whenever we read one of those special light textures. Problem solved?

I tried this approach and it looks the same to me, consistent with the complex cancellng system that we had before. Also fixes the Perseus glass and Procyon star chart issues addressed in #1404.

An explanation for the canceling approach is given in #1050:
> We cannot overBright separate stage as stage are clamped within [0.0, 1.0] when blending multiple stages together, meaning the result of the multiplication of a separate light stage with a light factor will be clamped before blending it with the color map.

But this is not an accurate representation of how shaders work. Each stage acts independently and directly on the color and depth buffers. There is no scratch area where shader stages are blended before going into the color buffer.

I would expect this to ameliorate the color banding problems with low-precision color buffers since now we don't give up 2 bits of precision to make room for canceling.

I haven't tried vertex lighting yet. Not sure how it's supposed to work with overbright.

DolceTriade commented:
Screenies?

VReaperV commented:
> I haven't tried vertex lighting yet. Not sure how it's supposed to work with overbright.

Vertex lighting seems to use inverse light factor the same way.

slipher commented:
From a recent master, default settings:
![unvanquished-atcshd-outside](https://github.com/user-attachments/assets/d8dc9b8e-fac5-4d39-93ba-77d70107a3f8)
![unvanquished-nano-overmind](https://github.com/user-attachments/assets/e28757ac-a837-4710-b047-c0d7fc9ed90a)

With this PR:
![unvanquished-atcshd-outside](https://github.com/user-attachments/assets/1803cb07-1e12-42ce-8fb3-af472f21f9e5)
![unvanquished-nano-overmind](https://github.com/user-attachments/assets/2444daf6-41fd-4881-a170-e895469189b1)

illwieckz commented:
I'm all in favor of a simpler implementation.

In the past I first tried what you did there, but I got bugs. Maybe they were mistakes I did?

My main concern was about surfaces where the lightmap is applied more than once (and then multiplied more than once), basically terrain blended surfaces. Fortunately with your branch those terrain blended surfaces look fine (thunder outdoor and station15 garden):

![unvanquished_2024-10-31_035141_000](https://github.com/user-attachments/assets/2e71d17e-42e2-439f-95f4-09e4b2c10844)

![unvanquished_2024-10-31_034606_000](https://github.com/user-attachments/assets/ee8243aa-967c-4e42-b2b4-4bf7e04b5709)

On the station15 garden, plant leaves are wrongly lit, maybe they are vertex lit.

So, here is the light sources that should be multiplied:

- lightmap light
- vertex light
- grid light
- white fullbright light WHEN the lightmode is not fullbright

illwieckz commented:
```
> - white fullbright light WHEN the lightmode is not fullbright
```

Doing this properly is one of the fix from #1404 (the starchart issue) and it should be done in all implementations.

illwieckz commented on src/engine/renderer/glsl_source/computeLight_fp.glsl:
```diff
@@ -45,6 +47,8 @@ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 		ambientColor = ambientScale * texel.rgb;
 		lightColor = directedScale * texel.rgb;
+		ambientColor /= u_InverseLightFactor;
+		lightColor /= u_InverseLightFactor;
```
Better do some `something *= u_LightFactor` instead.

illwieckz replied:
I added some commits to do that.

illwieckz commented:
On my end when I first tried your approach, I was having such places reproducing the same clamping as we get whih the legacy clamped overbright:

![unvanquished_2024-10-31_041056_000](https://github.com/user-attachments/assets/af020d0b-f232-4e6b-b60e-d1dcb431de0f)

But as you see in this screenshot, your branch doesn't reproduce the clamping. That clamping in GLSL was also what prompted me to follow a more complex route to avoid that clamping.

I don't know why I got any clamping in the past even when doing overbright in GLSL that way. The GLSL code received heavy rewrites, maybe we fixed some bugs in the meantime?

slipher commented:
> So, here are the light sources that should be multiplied:
> 
>     * lightmap light
> 
>     * vertex light
> 
>     * grid light
> 
>     * white fullbright light WHEN the lightmode is not fullbright

Lightmap and grid light are what is implemented in the initial draft of this PR. I don't think the last case exists for the simplified approach; it sounds like a problem of canceling.

Vertex light is a tiny bit more complex than the other cases since everything has the vertex color attributes, not only things with precomputed lighting. Maybe I can check `tess.bspSurface` and scale u_ColorModulate by 4 if is true.

illwieckz commented:
Here is an example of GLSL overbrigth clamping when I first applied the overbright on light per stage instead of screen:

![unvanquished_2024-02-06_175946_000](https://github.com/user-attachments/assets/b61ddcad-a512-457b-af5e-3d402cc2f132)

I don't get why I got it and you're not getting it, but if we're not getting it with the per-light approach, that's great!

illwieckz commented:
> I don't think the last case exists for the simplified approach; it sounds like a problem of canceling.

Probably not, the fullbright light mode is to just apply a white lightmap, so it's just using the lightmap code. Maybe there is nothing to do.

Though, we must make sure the (multiplied) vertex light code doesn't multiple the (multiplied) white lightmap fallback.

illwieckz commented:
I reproduce the clamped overbright with that GLSL implementation, now I remember when it's happening.

The clamped overbright happens with such GLSL implementation when the lightmap is blended as a separate stage (i.e. no collapsing is doable), when the engine processes a lightmap stage over a colormap stage (or the colormap stage over a lightmap stage), instead of a singe diffusemap stage:

This:

```
stage colormap
stage lightmap
```

Or this:

```
stage lightmap
stage colormap
```

Instead of:

```
stage diffusemap
```

Here is an example with a build of the pulse map attached (the one hosted on sweet folder has broken lightstyles so it doesn't reproduce the light clamping bug):

The `master` branch:

![unvanquished_2024-10-31_043029_000](https://github.com/user-attachments/assets/525322db-55ac-4acc-8039-c1ede8b6c66e)

This branch:

![unvanquished_2024-10-31_044425_000](https://github.com/user-attachments/assets/efcf8f4c-664a-4542-b7a6-7bdabd20d530)

File: [map-pulse_0-20230207-114641-e9a0a5e.dpk.zip](https://github.com/user-attachments/files/17581693/map-pulse_0-20230207-114641-e9a0a5e.dpk.zip)

I guess I don't reproduce the bug with plat23 anymore because I improved the collapse code and we now collapse more materials, getting them processed as a single diffusemap stage instead of separate colormap + lightmap stages.

illwieckz commented:
To make it more obvious, the problem is where the red arrows are pointing at:

![unvanquished_2024-10-31_044425_001](https://github.com/user-attachments/assets/3d2c4177-68cd-4352-b593-75fc549d58d5)

illwieckz commented:
So I actually reproduce the bug with a released Tremulous map that is the original map build by the original author (only repackaged, not rebuilt):

Map: https://users.unvanquished.net/~sweet/pkg/map-metro-b1-2_0%2B0.dpk

viewpos: `2123 -231 -166 -89 26`

With `master`:

![unvanquished_2024-10-31_051304_000](https://github.com/user-attachments/assets/652dc0f0-5bf7-4ff5-b1fe-489812d7567f)

With this branch:

![unvanquished_2024-10-31_051423_000](https://github.com/user-attachments/assets/9228a029-7998-4dbb-af87-346f3c7b10b2)

illwieckz commented:
Interestingly, your branch fixes the (what is assumed to be a) floating light shader that should not be visible:

Map: https://users.unvanquished.net/~sweet/pkg/map-metro-b1-2_0%2B0.dpk

viewpos; `2267 -429 -223 -51 -19`

branch `master` or `illwieckz/obb`:

![unvanquished_2024-10-31_052550_000](https://github.com/user-attachments/assets/e5d11009-44c6-446f-8166-734549f0a943)

branch `slipher/overbright-simple`:

![unvanquished_2024-10-31_052655_000](https://github.com/user-attachments/assets/558e1c38-3b14-4af8-be3c-0ad917757678)

illwieckz commented:
So basically, I'm all in favor of investing this more and if we can make it handle light styles properly, let's do it.

But for the 0.55.1 release we better merge my fix, we'll have all the time to investigate a different implementation after that.

slipher commented:
I looked at the Metro light style shaders. So apparently it builds a composite lightmap in the color buffer from 4 additive stages, then does a standard multiplicative blend with the diffusemap in the final stage.

I guess we could keep the cancellation approach for light styles only. There is already detection and special handling for light styles anyway.

illwieckz commented:
> I guess we could keep the cancellation approach for light styles only. There is already detection and special handling for light styles anyway.

If that's doable that would be awesome! I dislike the cancelling-everywhere approach a lot (it's very fragile because of the very large amounts of uses case…).

VReaperV commented:
> > So, here are the light sources that should be multiplied:
> > 
> > * lightmap light
> > * vertex light
> > * grid light
> > * white fullbright light WHEN the lightmode is not fullbright
> 
> Lightmap and grid light are what is implemented in the initial draft of this PR. I don't think the last case exists for the simplified approach; it sounds like a problem of canceling.
> 
> Vertex light is a tiny bit more complex than the other cases since everything has the vertex color attributes, not only things with precomputed lighting. Maybe I can check `tess.bspSurface` and scale u_ColorModulate by 4 if is true.

Vertex lighting is multiplied with the diffuse colours, so multiplying it by 4 again would be wrong. 

Although on master vertex lighting is also somewhat broken with overbright: e. g. on map `railgun` from Wolf:ET the icicles are almost completely black without overbright clamping, owing to the fact that they're entirely vertex-lit. Perhaps there needs to be a check for surfaces that use only vertex lighting, which would then be multiplied or whatever. That shouldn't be done in tess however, just multiply the vertex colours themselves instead.

illwieckz commented:
> Although on master vertex lighting is also somewhat broken with overbright: e. g. on map railgun from Wolf:ET the icicles are almost completely black without overbright clamping

Probably the same bug as the procyon hologram?

VReaperV commented:
> > Although on master vertex lighting is also somewhat broken with overbright: e. g. on map railgun from Wolf:ET the icicles are almost completely black without overbright clamping
> 
> Probably the same bug as the procyon hologram?

I doubt it, the procyon one has a lightmap.

slipher commented:
Besides canceling, another possible approach to light styles would be using multiple collapse lightmap stages with additive blending, and disable dynamic lights in all but one.

illwieckz commented:
> disable dynamic lights in all but one.

I guess it is already the case, at least with the tiled renderer: light styles light map are expected to be rendered with `Renger_generic3D` and not with `Render_lightMapping` as far as I know.

illwieckz commented:
> multiple collapse lightmap

Actually I don't see how that would works? to collapse lightmaps with what?

illwieckz commented:
This may be theoretical, but we may not exclude that some non-lightstyle lightmaps rendered with the lightMapping stage get blended like the light style ones and reproduce the bugs that happens with light styles.

It's just that the collapsing has been well improved so standalone non-light-style lightmap stages are more rare nowadays, but it is not guaranteed we collapse all of them or that we even can collapse all of them.

illwieckz commented:
@slipher do you have any understanding about how the light style lightmap blending is working and then how we may workaround the related bugs?

Actually we may workaround some wrongness by normalizing to `1` at the end of the GLSL shader, to avoid bad colors popping up there and there. But that will still clamp them (it would still be better than the legacy lightmap shift that is not actually clamping but turns 75%s of the data into garbage).

illwieckz commented:
I added a commit to multiply the light factor instead of dividing the inverse light factor, that also cleaned-up code, so it helps to have a better look at things.

VReaperV commented:
> Interestingly, your branch fixes the (what is assumed to be a) floating light shader that should not be visible:
> 
> Map: https://users.unvanquished.net/~sweet/pkg/map-metro-b1-2_0%2B0.dpk
> 
> viewpos; `2267 -429 -223 -51 -19`
> 
> branch `master` or `illwieckz/obb`:
> 
> ![unvanquished_2024-10-31_052550_000](https://private-user-images.githubusercontent.com/2311649/381824449-e5d11009-44c6-446f-8166-734549f0a943.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3MDUwMzgsIm5iZiI6MTczMDcwNDczOCwicGF0aCI6Ii8yMzExNjQ5LzM4MTgyNDQ0OS1lNWQxMTAwOS00NGM2LTQ0NmYtODE2Ni03MzQ1NDlmMGE5NDMuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMDRUMDcxODU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzE1Yzk1YTk4NDk2ZGFkOTliZmRlZDFmYjcxMjc2OTAyODZlMjQ5YWFkMDA1NDcxMzllZWY3MTY1YWQwMjExNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.WRuiAqBcS69GcwjCfzUISxidquZd4goKWRt1heHcsC8)
> 
> branch `slipher/overbright-simple`:
> 
> ![unvanquished_2024-10-31_052655_000](https://private-user-images.githubusercontent.com/2311649/381824454-558e1c38-3b14-4af8-be3c-0ad917757678.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3MDUwMzgsIm5iZiI6MTczMDcwNDczOCwicGF0aCI6Ii8yMzExNjQ5LzM4MTgyNDQ1NC01NThlMWMzOC0zYjE0LTRhZjgtYmUzYy0wYWQ5MTc3NTc2NzguanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMDRUMDcxODU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjIyODdkYmRjNDA5YWJlZTQ4N2YzOTU1NmMwZDYzOGZlYTU0ZDRmODZiZmJjMTgzNjY3ZjgzZjA2ZmZmZTVjNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.7Cx0Nh_s-BCYNs3tCBgOLzNPUfMEBncVvyR0NVeFahE)
> 
> The metro map is the only map known to reproduce this floating transparent surface bug, so it's low priority, but anything helping to pinpoint what is happening is welcome.

I'm guessing it's the weird square on the wall that is the issue, right?

illwieckz commented:
> I'm guessing it's the weird square on the wall that is the issue, right?

For the “floating light shader”, yes.

illwieckz commented:
I rebased on current master, fixed some merge conflict, and cleaned-up more cancelling leftovers.

@VReaperV I would love if you could look at that broken light style at said position in map with your magical tools and if you can actually describe what's going on and what could explain the light being that bad.

Maybe light styles are the only things that require a workaround when we multiply the lightmap directly in color/lightmapping shader.

slipher commented:
> I rebased on current master, fixed some merge conflict, and cleaned-up more cancelling leftovers.

@illwieckz I would suggest getting rid of your commit `fixup: only multiply generic3D with light factor if it's a light style lightmap`. This means treating the affected light map like as if overbright bits were configured to 0 (instead of 2).

VReaperV commented:
> I rebased on current master, fixed some merge conflict, and cleaned-up more cancelling leftovers.
> 
> @VReaperV I would love if you could look at that broken light style at said position in map with your magical tools and if you can actually describe what's going on and what could explain the light being that bad.
> 
> Maybe light styles are the only things that require a workaround when we multiply the lightmap directly in color/lightmapping shader.

Apparently it's some surface that has white image as both diffuse and lightmap, and is rendered using the `lightMapping` shader. It's also vertex-lit. On master at least, that is.

VReaperV commented:
> I rebased on current master, fixed some merge conflict, and cleaned-up more cancelling leftovers.
> 
> @VReaperV I would love if you could look at that broken light style at said position in map with your magical tools and if you can actually describe what's going on and what could explain the light being that bad.
> 
> Maybe light styles are the only things that require a workaround when we multiply the lightmap directly in color/lightmapping shader.

So the reason it doesn't show up on this branch is because the surface is rendered with blendfunc GL_DST_COLOR , GL_ZERO and GL_FUNC_ADD, and this branch produces a completely white fragment, while master doesn't.

The "before" value is essentially multiplied with the fragment value:
On master:
![image](https://github.com/user-attachments/assets/e4dfd164-550c-4828-99ee-9878e110bab0)
On this branch:
![image](https://github.com/user-attachments/assets/91654567-9896-4b6e-b997-91595c59e524)

While on master `u_InverseLightFactor` is set to -0.25 for this surface, which means it's not used and the fragment essentially gets the vertex colours, this branch sets `u_LightFactor` to 4.0 for this surface, and then the lightmap colour is multiplied by it, resulting in light colour of {4.0, 4.0, 4.0, 1.0}, which is then multiplied with the vertex colours, which in turn is enough to create a {1.0, 1.0, 1.0, 1.0} fragment.

illwieckz commented:
> So the reason it doesn't show up on this branch is because the surface is rendered with blendfunc GL_DST_COLOR , GL_ZERO and GL_FUNC_ADD, and this branch produces a completely white fragment, while master doesn't.

Hmm, is this an explanation for why the square light emitting shader is displayed with `master` branch?

![unvanquished_2024-10-31_052550_001](https://github.com/user-attachments/assets/f3183b62-a276-4d51-869b-700665ab1de1)

I was looking for an explanation for why the light style is broken with this branch, so that weirdly colored halo on the wall:

![unvanquished_2024-10-31_052655_001](https://github.com/user-attachments/assets/fd9ddae1-47e5-4920-9ebf-096a94612b4f)

I suspect some kind of clamping.

VReaperV commented:
> Hmm, is this an explanation for why the square light emitting shader is displayed with `master` branch?

Yep.

> I was looking for an explanation for why the light style is broken with this branch, so that weirdly colored halo on the wall:

...oh. I didn't realise that was the issue lol.

illwieckz commented:
That thread was a big confusing I agree. 😄️

Before comment https://github.com/DaemonEngine/Daemon/pull/1406#issuecomment-2454011193 I was talking about the light emitting shader, but starting with that one I was talking about light styles. 🙂️

illwieckz commented:
![unvanquished_2024-11-05_115510_000](https://github.com/user-attachments/assets/b6260921-f2ed-4015-8831-12e8bd8cf850)

🤔️

VReaperV commented:
That looks more like overdark to me.

slipher commented:
Dark map on material system fixed.

Trying a new approach where we configure the FBO to allow colors greater than 1. This also has side effects though...

![unvanquished_2024-11-05_051138_000](https://github.com/user-attachments/assets/fac8d3ec-249b-4ebc-82b2-6e86e0a9c142)

VReaperV commented:
Given that `u_LightFactor` is only used for lightmap and lightgrid, can't we just multiply those and avoid the shader hack altogether?

illwieckz commented:
> Given that `u_LightFactor` is only used for lightmap and lightgrid, can't we just multiply those and avoid the shader hack altogether?

This is the original implementation making it clamped…

VReaperV commented:
> > Given that `u_LightFactor` is only used for lightmap and lightgrid, can't we just multiply those and avoid the shader hack altogether?
> 
> This is the original implementation making it clamped…
> 
> Edit: lightmap, lightgrid, and vertex color.

We could try using a different image format there.

illwieckz commented:
I added two changes:

1. Also multiply the lightmaps used in light style.
2. Do not multiply fullbright light, I first believed this should also be multiplied, but it was not the case before.

The second change fixes the procyon hologram:

![unvanquished_2024-11-06_182740_000](https://github.com/user-attachments/assets/985f96c3-546c-47cb-97fb-6ba8e3e19692)

illwieckz commented:
The metro visible “visible light emitting shader” is a tricky one.

As a remind, the lantern, the candle, etc. come with some shaders having a single stage like this:

```
	{
		map $lightmap
		blendfunc filter
	}
```

So we get floating lightmaps multiplied over whatever is behind.

I even don't know why it worked on Tremulous…

Maybe we can just strip them out…

illwieckz commented:
I added a commit to fix the lantern/candle. I assume it's a shader bug, it doesn't make sense. That patch has nothing to do with overbright and could be done in another PR if we wanted to.

illwieckz commented:
I don't see any bugs anymore, in all maps I've tested (metro, procyon, pulse…)

illwieckz commented:
So, I noticed there was something still wrong with leaves in station15 garden.

Here is 0.54.1, the leaves on the floor are correct:

![unvanquished_2024-11-07_135251_000](https://github.com/user-attachments/assets/c5a3418f-8e15-4cc5-9a35-ff8d0d4a0cb2)

Here is this branch before we only multiplied light style/light map/diffuse map stages, the leaves on the floor are too dark:

![unvanquished_2024-11-07_135339_000](https://github.com/user-attachments/assets/d60204b3-15a2-45b7-bf73-6cee0d207ee9)

HEre is this branch when I multiplied all colormaps (which is wrong), just to verify something, the leaves on the floor are correct, but the ones in tree are too bright:

![unvanquished_2024-11-07_135416_000](https://github.com/user-attachments/assets/22c14d22-7dee-4c6c-8279-2fc1dd7b7aed)

Here is once I added the patch multiplying color map with lightFactor when rgbGen is `vertex`:

![unvanquished_2024-11-07_140242_000](https://github.com/user-attachments/assets/8df17ef4-aa61-436f-868d-720331817589)

So here is an example of such leaf-on-the-ground shader:

```
textures/station15_custom/leave02d
{
	qer_editorImage textures/station15_custom_src/leave02d_d
	qer_alphaFunc greater .5

	q3map_cloneShader textures/station15_custom/leave02d_back
	q3map_vertexScale 1.5

	surfaceparm alphashadow
	surfaceparm trans
	surfaceparm nonsolid
	surfaceparm nomarks

	{
		map textures/station15_custom_src/leave02d_d
		rgbGen vertex
		depthWrite
		alphaFunc GE128
	}
}
```

Such shader doesn't have a lightmap, so it's processed as a generic shader.

But the colormap is blended with `rgbGen vertex` meaning it's multiplied with the vertex light.

So this is an old trick to do single-stage vertex-based lighting.

And indeed the old overbright code was multiplying vertex colors at map load.

So we need to multiply the vertex color like if it is a light.

I don't know yet why the foam is too bright, though.

illwieckz commented:
I guess we should do something when using `rgbgen oneMinusVertex` because the vertex light was multiplied too before.

illwieckz commented:
Here is the foam shader:

```
gfx/station15/haze
{
	entityMergable
	cull none
	{
		map gfx/station15/water_haze
		blendFunc blend
		rgbGen vertex
		alphaGen vertex
	}
}
```

I don't know what to do to fix it.

illwieckz commented:
> I guess we should do something when using `rgbGen oneMinusVertex` because the vertex light was multiplied too before.
> 
> But I don't know what to do in such case.

I haven't found a single occurrence of `rgbGen oneMinusVertex` in the whole betaserv dump.

There are some `alphGen oneMinusVertex` in some maps like transit, but I can't find usage of `oneMinusVertex` with `rgbGen`.

illwieckz commented:
Maybe the haze always have been too bright but clamping hidden the bug.

illwieckz commented:
If I remove `rgbgen vertex` from the haze shader, it looks correct:

![unvanquished_2024-11-07_150056_000](https://github.com/user-attachments/assets/d3f3ebae-64d1-44ef-b018-98f8a4f190fd)

So I guess the previous clamped overbright just set the vertex color to `{1, 1, 1}`, making the texture fullbright, but that was wrong.

illwieckz commented:
It is possible to precompute the per-stage light factor at map loading time, but that would require to rewrite some of the BSP loading code (most shaders are loaded before the global light factor is known, and there are some tricky chicken-and-egg problem to solve).

illwieckz commented:
The haze effect is not a map surface, it's a particle emitted surface, so I guess it has no vertex color, and we may work this around.

illwieckz commented:
> The haze effect is not a map surface, it's a particle emitted surface, so I guess it has no vertex color, and we may work this around.

Now done:

![unvanquished_2024-11-07_162100_000](https://github.com/user-attachments/assets/9e83abcf-ed24-438e-a63d-0768d5f8eb39)

In fact such surface should better be collapsed as a diffuse map with vertex lighting, so they can also receive dynamic lights.

They are meant to receive light, it's just a trick to use vertex light explicitely instead of a lightmap, and the light map code already calls functions like `isExplicitelyVertexLitSurface()`.

Such collapse is to be postponed for the future.

illwieckz commented:
I added a commit to add `gl_generic3DShader`, this way every call of `gl_genericShader` don't need to receive any `lightFactor` (it no need to spam all the debug code to set it to 1). I have not stripped `gl_genericShader` from all the macros that are now useless, but we can do it.

illwieckz commented:
Once material system is fixed, this will be ready to merge.

VReaperV commented:
> @VReaperV I reported the changes to the material system, but I probably did a stupid mistake somewhere as most textures are black, can you look at it?

Yeah, will do in a bit.

VReaperV commented on src/engine/renderer/Material.cpp:
```diff
@@ -214,45 +214,45 @@ void UpdateSurfaceDataGeneric3D( uint32_t* materials, Material& material, drawSu
-	float inverseLightFactor = pStage->cancelOverBright ? tr.mapInverseLightFactor : 1.0f;
-	gl_genericShaderMaterial->SetUniform_InverseLightFactor( inverseLightFactor );
+	// u_LightFactor
+	float lightFactor = ( pStage->type == stageType_t::ST_STYLELIGHTMAP || pStage->type == stageType_t::ST_STYLECOLORMAP || ( tess.bspSurface && pStage->rgbGen == colorGen_t::CGEN_VERTEX ) ) ? tr.mapLightFactor : 1.0f;
```
Redundant parentheses.

VReaperV replied:
Use `drawSurf->bspSurface` here, `tess.bspSurface` won't be valid in this functions.

VReaperV commented on src/engine/renderer/gl_shader.h:
```diff
@@ -2140,7 +2140,7 @@ class u_LightFactor :
 public:
 	u_LightFactor( GLShader *shader ) :
-		GLUniform1f( shader, "u_LightFactor" )
+		GLUniform1f( shader, "u_LightFactor", true )
```
If `u_LightFactor` can change between surfaces, then this should not be global.

VReaperV reviewed (commented):
Why the change to have 3 generic shaders instead of 2?

VReaperV commented:
I'll check now if fixing the above makes it work with material system.

VReaperV commented:
@illwieckz I'm not getting any black textures:
![image](https://github.com/user-attachments/assets/36536227-4c36-4a55-8b55-2d7a4e4ef78e)
Can you confirm if the issue is still there on your end?

illwieckz commented:
> Why the change to have 3 generic shaders instead of 2?

To avoid the need of calling `SetUniform_InverseLightFactor( 1.0 )` on every function using the generic shader (like debug tools), and the risk of forgetting about it.

illwieckz commented:
> Can you confirm if the issue is still there on your end?

I'll test this after sleeping. 🙂️

slipher commented:
I'd like to avoid adding u_LightFactor to the generic shader at all. I think we can do it via u_Color instead. We can add a flag to make Tess_ComputeColor scale by the light factor. I'll try this later.

illwieckz commented:
I still have one bug with the material system, off:

![unvanquished_2024-11-09_134108_000](https://github.com/user-attachments/assets/b080820f-3fd7-4bdf-b382-178ec3531728)

on:

![unvanquished_2024-11-09_134213_000](https://github.com/user-attachments/assets/01e4690a-14a0-4598-a579-ceeb28dbed77)

You'll see the plants are too bright.

(Don't mind if the floor and the trees look a bit different than stock map, I'm also experimenting other things with the map, but those plants are not modified).

VReaperV commented:
It's because the vertex colour attribute is disabled. I'll look at why that happens.

VReaperV commented:
> I still have one bug with the material system, off:
> 
> ![unvanquished_2024-11-09_134108_000](https://private-user-images.githubusercontent.com/2311649/384606761-b080820f-3fd7-4bdf-b382-178ec3531728.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzExNTkwNDcsIm5iZiI6MTczMTE1ODc0NywicGF0aCI6Ii8yMzExNjQ5LzM4NDYwNjc2MS1iMDgwODIwZi0zZmQ3LTRiZGYtYjM4Mi0xNzhlYzM1MzE3MjguanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMDlUMTMyNTQ3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDA5MWU0YWQxYmFmMDk0YjI2NzYyMmYyNjQ5MGYzNWYzYWU4MGZiNTA5ZTdkYjAxMzQ1ZDYxZmNkMzNmMzQwNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.K0nKAMSmUO3N_hNKjItCpfV_WO7GhFjlrmxFqsX_a1A)
> 
> on:
> 
> ![unvanquished_2024-11-09_134213_000](https://private-user-images.githubusercontent.com/2311649/384606775-01e4690a-14a0-4598-a579-ceeb28dbed77.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzExNTkwNDcsIm5iZiI6MTczMTE1ODc0NywicGF0aCI6Ii8yMzExNjQ5LzM4NDYwNjc3NS0wMWU0NjkwYS0xNGEwLTQ1OTgtYTU3OS1jZWViMjhkYmVkNzcuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMDlUMTMyNTQ3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGIyNjQ1ZjdiOTY0NGI3ZGFiMWRjYmM2OWVhODBkZDAzODI4OTE1YzZjMmRlOGE3OWQ0M2Q2YzU3MWM5NzVjMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.zc1RE2EGfeukfl9e3eioPJxL4Ka-PDMS8dzqlc2NFwo)
> 
> You'll see the plants are too bright.
> 
> (Don't mind if the floor and the trees look a bit different than stock map, I'm also experimenting other things with the map, but those plants are not modified).

Fixed now.

VReaperV reviewed (approved):
LGTM

VReaperV commented:
Since I only did a small change here and the pool of people who can review this isn't much larger than this pr's contributors, I took the liberty of reviewing it, especially as I reviewed most of it earlier anyway.

illwieckz commented:
I squashed the “fixup” commits.

This looks ready to me.

There are some improvements that may be doable:

- setting per-stage lightFactor at shader load time (this requires to rewrite some BSP loading code),
- pre-multiplying lightFactor with u_Color to not have to pass it to GLSL (this may benefit from the former),
- turn colormap stage with vertex color blending to a diffusemap one when the shader has no “nolightmap” keyword (so it can also receive dynamic light).

But that can be done in the future in subsequent PRs.

illwieckz commented:
I moved the fix for the metro lanterns here:

- https://github.com/DaemonEngine/Daemon/pull/1426

illwieckz commented:
I also renamed the PR to fit what it does, as it is now in good shape for merging and we will going to do things that way.

slipher commented:
Please remember that according to our PR etiquette, no one should merge another person's PR :)

slipher commented:
I've now reimplemented the three special cases treated by illwieckz in way that avoids adding `u_LightFactor` to the generic shader. Those fixes are separated into a temporary fixup commit for convenient reading.

Let us call BASE the source tree obtained from merging #1426, #1431 and #1434 into master. Let us call SIMPLE this PR's tree, the `slipher/overbright/simple` branch (which incorporates the three aforementioned PRs). To verify the changes I have, using the scenes listed [here](https://github.com/slipher/imgdiff/blob/master/imgdiff.cfg), done a four-way comparison of BASE with all default settings, BASE with material system, SIMPLE with all default settings, and SIMPLE with material system. (Please suggest more scenes!)

## Some bugs fixed by these overbright changes

### Non-lightmapped vertex-lit surface not overbright scaled (station15 `462 -1428 -114 178 -8`)
BASE

![unvanquished-station15-vertexlit-veg](https://github.com/user-attachments/assets/8334455b-28af-4a84-a2e8-0f1976d2595e)

SIMPLE
![unvanquished-station15-vertexlit-veg](https://github.com/user-attachments/assets/a0f87fc3-46ba-4c45-bdb2-6aa2937218db)

### Improperly scaled 'glow map' (zittrig-arena `1201 -301 -338 87 0`)
BASE ( + material system)
![unvanquished_2024-11-12_010216_000](https://github.com/user-attachments/assets/523e7ceb-3147-420a-a020-a2627f4b4662)

SIMPLE ( + material system)
![unvanquished_2024-11-12_010340_000](https://github.com/user-attachments/assets/a6805270-3da1-44a3-bccc-a99a8e6d7316)

### Procyon star chart darkened with material system
BASE + material system

![unvanquished-procyon-starchart](https://github.com/user-attachments/assets/d698720b-bfe2-48db-91a2-50fe7093267e)

SIMPLE + material system
![unvanquished-procyon-starchart](https://github.com/user-attachments/assets/0cb42945-44a4-4aab-8b08-1fef1acd7302)

## Consequences for framebuffers with only 1 byte of color

In some cases things are worse because of clamping in the color buffer that wasn't there before. In some cases things are a bit better because of the effective increased precision. It may be preferable to enable overbright clamping on low-end systems that don't have 16-bit float framebuffers.

### Color clamping regression - Pulse `4822 -1312 -1314 98 13`
BASE + `r_highPrecisionRendering 0`
![unvanquished-pulse-lowprecision-clamping](https://github.com/user-attachments/assets/9ecc9ecf-da66-4c9d-b57b-8f95e3384eac)

SIMPLE + `r_highPrecisionRendering 0`
![unvanquished-pulse-lowprecision-clamping](https://github.com/user-attachments/assets/b71f40a8-c1df-46e5-9863-d35cae0a7af4)

### Decreased noise in dark area - Procyon `-955 3443 276 94 7`
BASE + `r_highPrecisionRendering 0`
![unvanquished-procyon-lowprecision-noise](https://github.com/user-attachments/assets/9b9da893-6e96-4238-a7e1-0ba093f01843)

SIMPLE + `r_highPrecisionRendering 0`
![unvanquished-procyon-lowprecision-noise](https://github.com/user-attachments/assets/6ce9f5e4-945a-4352-9ae9-e3a95d195e4a)

illwieckz commented on src/engine/renderer/Material.cpp:
```diff
@@ -222,15 +222,12 @@ void UpdateSurfaceDataGeneric3D( uint32_t* materials, Material& material, drawSu
 
-	gl_genericShaderMaterial->SetUniform_ColorModulate( rgbGen, alphaGen );
+	bool mayUseVertexOverbright = pStage->type == stageType_t::ST_COLORMAP && drawSurf->bspSurface;
+	gl_genericShaderMaterial->SetUniform_ColorModulate( rgbGen, alphaGen, mayUseVertexOverbright );
```
What is preventing to multiply `tess.svars.color` in `Tess_ComputeColor`?

Why is the material variant calling ``Tess_ComputeColor` but not the non-material one? Is that a dormant bug we missed?

VReaperV replied:
Non-material code does `Tess_ComputeColor()` in `Tess_StageIteratorColor()`. Material system does not have that.

illwieckz replied:
Ah so in non-material code, `Tess_ComputeColor()` is done before calling `Render_generic3D()`. Is it too early to multiply the light factor? Because both `tess` and `pStage` are known as well.

VReaperV replied:
> Ah so in non-material code, `Tess_ComputeColor()` is done before calling `Render_generic3D()`. Is it too early to multiply the light factor? Because both `tess` and `pStage` are known as well.

I suppose that was just to avoid having to add it to all relevant stage renderers.

illwieckz reviewed (approved):
LGTM.
